mypy/checker.py

```diff
@@ -265,119 +265,125 @@ def check_overlapping_overloads(self, defn: OverloadedFuncDef) -> None:
-    #   (it's named `ts` after `send()`, since `tr` is `return`).
-    # - tr is the return type (the type of r in `return r`)
+    # - ty is the Yield type (the type of y in `yield y`)
+    # - tc is the type reCeived by yield (the type of c in `c = yield`).
+    # - tr is the Return type (the type of r in `return r`)
     #
     # A classic generator must define a return type that's either
-    # `Generator[ty, ts, tr]`, Iterator[ty], or Iterable[ty] (or
-    # object or Any).  If ts/tr are not given, both are Void.
+    # `Generator[ty, tc, tr]`, Iterator[ty], or Iterable[ty] (or
+    # object or Any).  If tc/tr are not given, both are Void.
     #
     # A coroutine must define a return type corresponding to tr; the
     # other two are unconstrained.  The "external" return type (seen
     # by the caller) is Awaitable[tr].
     #
+    # In addition, there's the synthetic type AwaitableGenerator: it
+    # inherits from both Awaitable and Generator and can be used both
+    # in `yield from` and in `await`.  This type is set automatically
+    # for functions decorated with `@types.coroutine` or
+    # `@asyncio.coroutine`.  Its single parameter corresponds to tr.
+    #
     # There are several useful methods, each taking a type t and a
     # flag c indicating whether it's for a generator or coroutine:
     #
     # - is_generator_return_type(t, c) returns whether t is a Generator,
-    #   Iterator, Iterable (if not c), or Awaitable (if c).
+    #   Iterator, Iterable (if not c), or Awaitable (if c), or
+    #   AwaitableGenerator (regardless of c).
     # - get_generator_yield_type(t, c) returns ty.
-    # - get_generator_receive_type(t, c) returns ts.
+    # - get_generator_receive_type(t, c) returns tc.
     # - get_generator_return_type(t, c) returns tr.
 
     def is_generator_return_type(self, typ: Type, is_coroutine: bool) -> bool:
         """Is `typ` a valid type for a generator/coroutine?
 
-        True if either Generator or Awaitable is a supertype of `typ`.
+        True if `typ` is a *supertype* of Generator or Awaitable.
+        Also true it it's *exactly* AwaitableGenerator (modulo type parameters).
         """
         if is_coroutine:
+            # This means we're in Python 3.5 or later.
             at = self.named_generic_type('typing.Awaitable', [AnyType()])
-            return is_subtype(at, typ)
+            if is_subtype(at, typ):
+                return True
         else:
             gt = self.named_generic_type('typing.Generator', [AnyType(), AnyType(), AnyType()])
-            return is_subtype(gt, typ)
+            if is_subtype(gt, typ):
+                return True
+        # Check that AwaitableGenerator exists -- it isn't defined in Python 2:
+        try:
+            self.lookup_qualified('typing.AwaitableGenerator')
+        except KeyError:
+            return False
+        agt = self.named_generic_type('typing.AwaitableGenerator',
```

Comment:
Wouldn't it be simpler to just look at the fullname instead, like we do elsewhere?